File: Homework-4/an_nguyen_task2.scala
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession
import scala.collection.mutable
import scala.collection.mutable.ListBuffer
import java.io.{File, PrintWriter}


object an_nguyen_task2 {
	def main(args: Array[String]): Unit = {
		val inputFile = args(0)
		val betweennessOutputFilePath = args(1)
		val communityOutputFilePath = args(2)
		Logger.getLogger("org").setLevel(Level.ERROR)
		val spark = SparkSession.builder.master("local[*]").appName("HW3").getOrCreate()

		val sc = spark.sparkContext
		val textRDD = sc.textFile(inputFile).map(x => x.split(" ")).map(x => (x(0), x(1)))
		val edges = textRDD.collect()
		val graph = mutable.Map[String, mutable.Set[String]]()
		for (pair <- edges) {
			if (!graph.contains(pair._1))
				graph += (pair._1 -> mutable.Set[String]())
			graph(pair._1).add(pair._2)
			if (!graph.contains(pair._2))
				graph += (pair._2 -> mutable.Set[String]())
			graph(pair._2).add(pair._1)
		}
		val vertices = mutable.Set[String]()
		for (i <- textRDD.map(x => x._1).collect())
			vertices += i
		for (j <- textRDD.map(x => x._2).collect())
			vertices += j

		// Calculate Betweenness
		val betweenness = calculateBetweenness(vertices, graph)
		val betweennessResult = betweenness.toList.map(x => (x._1._1, x._1._2, x._2)).sortBy(x => (-x._3, x._1, x._2))
		var output = betweennessOutputFilePath
		if (!betweennessOutputFilePath.contains("an_nguyen_task2_edge_betweenness_scala.txt"))
			output += "/an_nguyen_task2_edge_betweenness_scala.txt"
		var pw = new PrintWriter(new File(output))
		for (b <- betweennessResult)
			pw.write("('" + b._1 + "', '" + b._2 + "'), " + b._3 + "\n")
		pw.close()

		// Community Detection
		var result = communityDetection(vertices, graph, betweenness)
		result = result.sortBy(x => (x.length, x.head))
		output = betweennessOutputFilePath
		if (!communityOutputFilePath.contains("an_nguyen_task2_community_scala.txt"))
			output += "/an_nguyen_task2_community_scala.txt"
		pw = new PrintWriter(new File(output))
		for (c <- result)
			for (i <- c) {
				pw.write("'" + i + "'")
				if (c.indexOf(i) == c.length - 1)
					pw.write("\n")
				else
					pw.write(",")
			}
		pw.close()
	}

	def calculateBetweenness(vertices: mutable.Set[String], graph: mutable.Map[String, mutable.Set[String]]): mutable.Map[(String, String), Double] = {
		val betweenness = mutable.Map[(String, String), Double]()
		for (vertex <- vertices) {
			val bfs_result = bfs(vertex, graph)
			val edgeCredit = calculateEdgeCredit(bfs_result._1, bfs_result._2, bfs_result._3)
			for (edge <- edgeCredit.keys) {
				if (!betweenness.contains(edge))
					betweenness(edge) = 0
				betweenness(edge) += edgeCredit(edge) / 2
			}
		}
		betweenness
	}

	def bfs(root: String, graph: mutable.Map[String, mutable.Set[String]]): (ListBuffer[String], mutable.Map[String, ListBuffer[String]], mutable.Map[String, Int]) = {
		val queue = ListBuffer[String]()
		queue.append(root)
		val visited = ListBuffer[String]()
		visited.append(root)
		val level = mutable.Map[String, Int]()
		val parents = mutable.Map[String, ListBuffer[String]]()
		val shortestPath = mutable.Map[String, Int]()

		for (vertex <- graph.keys) {
			level(vertex) = -1
			parents(vertex) = ListBuffer[String]()
			shortestPath(vertex) = 0
		}
		shortestPath(root) = 1
		level(root) = 0

		while (queue.nonEmpty) {
			val current = queue.remove(0)
			visited.append(current)
			val children = graph(current)
			for (child <- children) {
				if (level(child) == -1) {
					queue.append(child)
					level(child) = level(current) + 1
				}
				if (level(child) == level(current) + 1) {
					parents(child).append(current)
					shortestPath(child) = shortestPath(child) + shortestPath(current)
				}
			}
		}
		(visited, parents, shortestPath)
	}

	def calculateEdgeCredit(visited: ListBuffer[String], parents: mutable.Map[String, ListBuffer[String]], shortestPath: mutable.Map[String, Int]): mutable.Map[(String, String), Double] = {
		val edgeCredit = mutable.Map[(String, String), Double]()
		val vertexCredit = mutable.Map[String, Double]()
		for (vertex <- visited)
			vertexCredit(vertex) = 1
		for (vertex <- visited.reverse)
			for (parent <- parents(vertex)) {
				val credit = vertexCredit(vertex) * shortestPath(parent) / shortestPath(vertex)
				val pair = ListBuffer[String]()
				pair.append(vertex)
				pair.append(parent)
				val orderedPair = Tuple2(pair.min, pair.max)
				if (!edgeCredit.contains(orderedPair))
					edgeCredit(orderedPair) = 0.0
				edgeCredit(orderedPair) += credit
				vertexCredit(parent) += credit
			}
		edgeCredit
	}

	def communityDetection(vertices: mutable.Set[String], graph: mutable.Map[String, mutable.Set[String]], betweenness: mutable.Map[(String, String), Double]): ListBuffer[ListBuffer[String]] = {
		val numberOfEdges = betweenness.size
		var btw = mutable.Map[(String, String), Double]() ++ betweenness
		val newGraph = mutable.Map[String, mutable.Set[String]]() ++ graph

		var maxModularity = -1.0
		var maxCommunities = ListBuffer[ListBuffer[String]]()
		while (btw.nonEmpty) {
			val communities = createCommunities(newGraph, vertices)
			val modularity = calculateModularity(communities, graph, numberOfEdges)
			if (modularity > maxModularity) {
				maxModularity = modularity
				maxCommunities = communities
			}
			val maxBetweenness: Double = btw.values.max
			var removedEdges = ListBuffer[(String, String)]()
			for (i <- btw)
				if (i._2 == maxBetweenness)
					removedEdges += i._1
			for (edge <- removedEdges) {
				newGraph(edge._1) = newGraph(edge._1).filter(_ != edge._2)
				newGraph(edge._2) = newGraph(edge._2).filter(_ != edge._1)
			}
			btw = calculateBetweenness(vertices, newGraph)
		}
		maxCommunities
	}

	def createCommunities(graph: mutable.Map[String, mutable.Set[String]], vertices: mutable.Set[String]): ListBuffer[ListBuffer[String]] = {
		var availableVertices = mutable.Set[String]()
		for (vertex <- vertices)
			availableVertices += vertex
		val communities = ListBuffer[ListBuffer[String]]()
		while (availableVertices.nonEmpty) {
			val community = createCommunity(graph, availableVertices.toList.head)
			val communityListBuffer = ListBuffer[String]()
			for (i <- community)
				communityListBuffer += i
			communities.append(communityListBuffer.sorted)
			availableVertices = availableVertices.diff(community)
		}
		communities
	}

	def createCommunity(graph: mutable.Map[String, mutable.Set[String]], vertex: String): mutable.Set[String] = {
		val community = mutable.Set[String]()
		val queue = ListBuffer[String]()
		queue.append(vertex)
		while (queue.nonEmpty) {
			val current = queue.remove(0)
			community.add(current)
			for (neighbor <- graph(current))
				if (!community.contains(neighbor))
					queue.append(neighbor)
		}
		community
	}

	def calculateModularity(communities: ListBuffer[ListBuffer[String]], graph: mutable.Map[String, mutable.Set[String]], numberOfEdges: Int): Double = {
		var totalModularity = 0.0
		for (community <- communities)
			for (i <- community)
				for (j <- community) {
					val k_i = graph(i).size
					val k_j = graph(j).size
					var a_ij = 0.0
					if (graph(i).contains(j))
						a_ij = 1
					totalModularity += a_ij - k_i * k_j / (2 * numberOfEdges.toFloat)
				}
		totalModularity / (2 * numberOfEdges.toFloat)
	}
}

File: Homework-4/an_nguyen_task2.py
import sys
import time
import operator
from pyspark.sql import SparkSession


def bfs(root, graph):
    queue = [root]
    visited = [root]
    level = {}
    parents = {}
    shortest_path = {}

    for vertex in graph.keys():
        level[vertex] = -1
        parents[vertex] = []
        shortest_path[vertex] = 0
    shortest_path[root] = 1
    level[root] = 0

    while queue:
        current = queue.pop(0)
        visited.append(current)
        children = graph[current]
        for child in children:
            if level[child] == -1:
                queue.append(child)
                level[child] = level[current] + 1
            if level[child] == level[current] + 1:
                parents[child].append(current)
                shortest_path[child] = shortest_path[child] + shortest_path[current]

    return visited, parents, shortest_path


def calculate_edge_credit(visited, parents, shortest_path):
    edge_credit = {}
    vertex_credit = {vertex: 1 for vertex in visited}
    for vertex in visited[::-1]:
        for parent in parents[vertex]:
            pair = min(vertex, parent), max(vertex, parent)
            credit = vertex_credit[vertex] * shortest_path[parent] / shortest_path[vertex]
            if pair not in edge_credit.keys():
                edge_credit[pair] = 0
            edge_credit[pair] += credit
            vertex_credit[parent] += credit
    return edge_credit


def calculate_betweenness(vertices, graph):
    betweenness = {}
    for vertex in vertices:
        visited, parents, shortest_path = bfs(vertex, graph)
        edge_credit = calculate_edge_credit(visited, parents, shortest_path)
        for edge, credit in edge_credit.items():
            if edge not in betweenness.keys():
                betweenness[edge] = 0
            betweenness[edge] += credit / 2
    return sorted(betweenness.items(), key=operator.itemgetter(1), reverse=True)


def create_community(graph, vertex):
    community = set()
    queue = [vertex]
    while queue:
        current = queue.pop(0)
        community.add(current)
        for neighbor in graph[current]:
            if neighbor not in community:
                queue.append(neighbor)
    return community


def create_communities(graph, vertices):
    temp = vertices.copy()
    communities = []
    while temp:
        community = create_community(graph, temp.pop())
        communities.append(sorted(list(community)))
        temp = temp.difference(community)
    return communities


def calculate_modularity(communities, graph, number_of_edges):
    total_modularity = 0.0
    for community in communities:
        for i in community:
            for j in community:
                k_i = len(graph[i])
                k_j = len(graph[j])
                a_ij = 1.0 if j in graph[i] else 0.0
                total_modularity += a_ij - k_i * k_j / (2 * number_of_edges)
    return total_modularity / (2 * number_of_edges)


def community_detection(vertices, graph, betweenness):
    number_of_edges = len(betweenness)
    btw, new_graph = betweenness.copy(), graph.copy()
    max_modularity, max_communities = -1, []
    while btw:
        communities = create_communities(new_graph, vertices)
        modularity = calculate_modularity(communities, graph, number_of_edges)
        if max_modularity < modularity:
            max_communities = communities
            max_modularity = modularity
        max_betweenness = max([v for _, v in btw])
        removed_edges = [edge for edge, value in btw if value == max_betweenness]
        for edge in removed_edges:
            new_graph[edge[0]] = {i for i in new_graph[edge[0]] if i != edge[1]}
            new_graph[edge[1]] = {i for i in new_graph[edge[1]] if i != edge[0]}
        btw = calculate_betweenness(vertices, new_graph)
    return max_communities


def main():
    start = time.time()

    input_filepath, betweenness_output_filepath, community_output_filepath = sys.argv[1:]
    spark = SparkSession.builder.master("local[*]").appName("HW4-Task2").getOrCreate()
    sc = spark.sparkContext
    sc.setLogLevel("ERROR")

    text_rdd = sc.textFile(input_filepath).map(lambda x: x.split(" ")).map(lambda x: (x[0], x[1]))
    edges = text_rdd.collect()
    graph = {}
    for pair in edges:
        if pair[0] not in graph.keys():
            graph[pair[0]] = set()
        graph[pair[0]].add(pair[1])
        if pair[1] not in graph.keys():
            graph[pair[1]] = set()
        graph[pair[1]].add(pair[0])
    vertices = set(text_rdd.map(lambda x: x[0]).collect() + text_rdd.map(lambda x: x[1]).collect())

    # Calculate Betweenness
    betweenness = calculate_betweenness(vertices, graph)
    with open(betweenness_output_filepath, "w") as f:
        for i in betweenness:
            f.write(f"{i[0]}, {i[1]}\n")

    # Community Detection
    result = community_detection(vertices, graph, betweenness)
    result.sort(key=lambda x: (len(x), x[0]))
    with open(community_output_filepath, "w") as f:
        for i in result:
            f.write("'" + "','".join(i) + "'\n")

    spark.stop()
    print(f"Duration: {time.time() - start}")


main()
